syntax = "proto3";

package grpc_modules.posts;
import "google/protobuf/timestamp.proto";
service PostService {
    rpc CreatePost (CreatePostRequest) returns (PostResponse);
    rpc GetPost (GetPostRequest) returns (PostResponse);
    rpc UpdatePost (UpdatePostRequest) returns (PostResponse);
    rpc DeletePost (DeletePostRequest) returns (EmptyResponse);
    rpc ListPosts (ListPostsRequest) returns (ListPostsResponse);
}

message CreatePostRequest {
    string title = 1;
    string description = 2;
    string creator_id = 3;
    bool is_private = 4;
    repeated string tags = 5;
}

message GetPostRequest {
    string id = 1;
    string user_id = 2;
}

message UpdatePostRequest {
    string id = 1;
    string user_id = 2;
    optional string title = 3;
    optional string description = 4;
    optional bool is_private = 5;
    repeated string tags = 6;
}

message DeletePostRequest {
    string id = 1;
    string user_id = 2;
}

message ListPostsRequest {
    string user_id = 1;
    int32 page = 2;
    int32 per_page = 3;
}

message PostResponse {
    string id = 1;
    string title = 2;
    string description = 3;
    string creator_id = 4;
    google.protobuf.Timestamp created_at = 5;
    google.protobuf.Timestamp updated_at = 6;
    bool is_private = 7;
    repeated string tags = 8;
}

message ListPostsResponse {
    repeated PostResponse posts = 1;
    int32 total = 2;
}

message EmptyResponse {}